### PR TITLE
Scheduler config validation

### PR DESCRIPTION
**Pull request by damemi**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Rebases/revives https://github.com/kubernetes/kubernetes/pull/68103, which added config validation for scheduler configs which specify invalid or contradictory combinations of policies

```release-note
NONE
```

xref: https://bugzilla.redhat.com/show_bug.cgi?id=1623346

### CODE REVIEW

File: cmd/kube-scheduler/app/server.go [hunk 1]
```diff
@@ -143,7 +144,7 @@ func runCommand(cmd *cobra.Command, args []string, opts *options.Options) error
 
 	// Apply algorithms based on feature gates.
 	// TODO: make configurable?
-	algorithmprovider.ApplyFeatureGates()
+	schedulerFeatureGateDependencies := algorithmprovider.ApplyFeatureGates()
```
Comment (ravisantoshgudimetla) [inline cmd/kube-scheduler/app/server.go]: Can we call this function `Run` directly or is there dependency in blocks further down in the function?

File: pkg/scheduler/algorithmprovider/defaults/defaults.go [hunk 2]
```diff
@@ -74,6 +77,11 @@ func ApplyFeatureGates() {
 		factory.RemovePredicateKeyFromAlgorithmProviderMap(predicates.CheckNodeDiskPressurePred)
 		factory.RemovePredicateKeyFromAlgorithmProviderMap(predicates.CheckNodePIDPressurePred)
 
+		excludedPredicates = sets.NewString(predicates.CheckNodeConditionPred,
```
Comment (ravisantoshgudimetla) [inline pkg/scheduler/algorithmprovider/defaults/defaults.go]: not against this logic but perhaps having a map[predicate.Name]=exclude and map[predicate.Name]=includedlist might make association much easier.

File: pkg/scheduler/api/validation/validation.go [hunk 3]
```diff
@@ -30,14 +30,51 @@ import (
...
+			}
+		}
+		for _, priority := range policy.Priorities {
+			if schedulerDependency.ExcludedPriorityList.Has(priority.Name) {
+				validationErrors = append(validationErrors, fmt.Errorf("Priority %s shouldn't be in when %s is enabled", priority.Name, schedulerDependency.Name))
+			}
+		}
+	}
+	for _, schedulerDependency := range featureDependencies {
+		for _, neededPredicate := range schedulerDependency.NeededPredicateList.List() {
```
Comment (ravisantoshgudimetla) [inline pkg/scheduler/api/validation/validation.go]: The 2 blocks can be converted to a single function.

File: pkg/scheduler/algorithmprovider/defaults/defaults.go (hunk 2 above)
Comment (damemi) [reply pkg/scheduler/algorithmprovider/defaults/defaults.go]: Well the underlying type for a Set is a map, and if you want to keep the `FeatureDependency` type as it is then we need the separate lists anyway. I also think having them separate is a little clearer

File: pkg/scheduler/api/validation/validation.go (hunk 3 above)
Comment (damemi) [reply pkg/scheduler/api/validation/validation.go]: Not really, one ranges over `policy.Predicates` and the other over `policy.Priorities` which are 2 different types (`[]PredicatePolicy` and `[]PriorityPolicy`) so I'm not sure if it's worth it for a small helper

File: cmd/kube-scheduler/app/server.go (hunk 1 above)
Comment (damemi) [reply cmd/kube-scheduler/app/server.go]: done

File: pkg/scheduler/algorithmprovider/defaults/defaults.go [hunk 4]
```diff
@@ -57,10 +58,12 @@ func defaultPredicates() sets.String {
 }
 
 // ApplyFeatureGates applies algorithm by feature gates.
-func ApplyFeatureGates() {
+func ApplyFeatureGates() []api.FeatureDependency {
```
Comment (ravisantoshgudimetla) [inline pkg/scheduler/algorithmprovider/defaults/defaults.go]: nit: Update the comment here as this function's signature changed.

File: pkg/scheduler/algorithmprovider/defaults/defaults.go (hunk 2 above)
Comment (ravisantoshgudimetla) [reply pkg/scheduler/algorithmprovider/defaults/defaults.go]: Fine with me but let's see if anyone else has any other opinions on it.

File: pkg/scheduler/algorithmprovider/defaults/defaults.go (hunk 4 above)
Comment (damemi) [reply pkg/scheduler/algorithmprovider/defaults/defaults.go]: done

**Review by ravisantoshgudimetla — APPROVED** [review]
/lgtm

/hold

waiting for @ahg-g to review this

/cc @Huang-Wei

**Comment by ravisantoshgudimetla** [comment]
Thanks for working on this @damemi :)

File: cmd/kube-scheduler/app/server.go [hunk 5]
```diff
@@ -179,6 +179,7 @@ func Run(cc schedulerserverconfig.CompletedConfig, stopCh <-chan struct{}) error
 		framework.NewRegistry(),
 		cc.ComponentConfig.Plugins,
 		cc.ComponentConfig.PluginConfig,
+		schedulerFeatureGateDependencies,
```
Comment (ravisantoshgudimetla) [inline cmd/kube-scheduler/app/server.go]: Thinking more about it, why can't we have it as part of `schedulerserverconfig`?

File: pkg/scheduler/api/validation/validation.go [hunk 6]
```diff
@@ -30,14 +30,51 @@ import (
 
 // ValidatePolicy checks for errors in the Config
 // It does not return early so that it can find as many errors as possible
-func ValidatePolicy(policy schedulerapi.Policy) error {
+func ValidatePolicy(policy schedulerapi.Policy, featureDependencies []schedulerapi.FeatureDependency) error {
```
Comment (liggitt) [inline pkg/scheduler/api/validation/validation.go]: this method should be called to see if a given config file is valid... it should generally be stateless, so it's not clear what the impact of the new second arg is. could that make a given config file sometimes valid and sometimes not?

Comment (damemi) [reply pkg/scheduler/api/validation/validation.go]: That's what this pull is trying to address: if a config file specifies certain policies that are technically valid but ultimately conflict based on what has been enabled or disabled through featuregates, then is that config actually valid? To answer your question, yes, this would make a given config sometimes valid and sometimes not depending on what features have been enabled for that instance of the scheduler.

File: cmd/kube-scheduler/app/server.go (hunk 5 above)
Comment (damemi) [reply cmd/kube-scheduler/app/server.go]: I tried this, but the versioned `KubeSchedulerConfig` type is hosted under `staging/kube-scheduler`, while we're adding `FeatureDependency` as a type in the `k/k/pkg/scheduler/api`. Currently `KubeSchedulerConfig` has no dependencies on that api and afaik `staging` can't depend on `k/k` right? So then this ends up breaking when you get to the internal/v1alpha1 conversions

Comment (damemi) [reply cmd/kube-scheduler/app/server.go]: If we made `FeatureDependency` a type under the `KubeSchedulerConfig` config api instead of pkg/scheduler/api, I think that would work. Do you want to do that instead?

File: pkg/scheduler/algorithmprovider/defaults/defaults.go [hunk 7]
```diff
@@ -84,7 +92,17 @@ func ApplyFeatureGates() {
 		factory.InsertPredicateKeyToAlgorithmProviderMap(predicates.PodToleratesNodeTaintsPred)
 		factory.InsertPredicateKeyToAlgorithmProviderMap(predicates.CheckNodeUnschedulablePred)
 
-		klog.Infof("TaintNodesByCondition is enabled, PodToleratesNodeTaints predicate is mandatory")
+		neededPredicates = sets.NewString(predicates.PodToleratesNodeTaintsPred)
```
Comment (ahg-g) [inline pkg/scheduler/algorithmprovider/defaults/defaults.go]: why wasn't CheckNodeUnschedulablePred added in the neededPredicates?

File: pkg/scheduler/api/types.go [hunk 8]
```diff
@@ -335,6 +336,15 @@ type HostPriority struct {
 	Score int
 }
 
+// FeatureDependency provides the feature and its dependent predicates and priorities.
+type FeatureDependency struct {
+	Name                  string      // Name of the feature gate.
+	NeededPredicateList   sets.String // List of predicates needed if we enable this predicate or priority
+	NeededPriorityList    sets.String // List of priorities needed if we enable this predicate or priority.
+	ExcludedPredicateList sets.String // List of predicates that should be excluded if we enable this predicate/priority.
+	ExcludedPriorityList  sets.String // List of priorities that should be excluded if we enable this predicate/priority
```
Comment (ahg-g) [inline pkg/scheduler/api/types.go]: The convention is to have the comment above the field, not on the same line.

File: pkg/scheduler/api/validation/validation.go [hunk 9]
```diff
@@ -30,14 +30,51 @@ import (
...
+				validationErrors = append(validationErrors, fmt.Errorf("Predicate %s shouldn't be present when %s is enabled", predicate.Name, schedulerDependency.Name))
+			}
+		}
+		for _, priority := range policy.Priorities {
+			if schedulerDependency.ExcludedPriorityList.Has(priority.Name) {
+				validationErrors = append(validationErrors, fmt.Errorf("Priority %s shouldn't be present when %s is enabled", priority.Name, schedulerDependency.Name))
+			}
+		}
+	}
+	for _, schedulerDependency := range featureDependencies {
```
Comment (ahg-g) [inline pkg/scheduler/api/validation/validation.go]: why not merge this for loop and the one in line 41?

File: pkg/scheduler/algorithmprovider/defaults/defaults.go (hunk 7 above)
Comment (damemi) [reply pkg/scheduler/algorithmprovider/defaults/defaults.go]: @ravisantoshgudimetla I pulled this from your commit, can you explain?

File: pkg/scheduler/api/types.go (hunk 8 above)
Comment (damemi) [reply pkg/scheduler/api/types.go]: fixed

File: pkg/scheduler/api/validation/validation.go (hunk 9 above)
Comment (damemi) [reply pkg/scheduler/api/validation/validation.go]: thanks, updated

File: pkg/scheduler/api/validation/validation.go (hunk 6 above)
Comment (damemi) [reply pkg/scheduler/api/validation/validation.go]: @liggitt bump, is that sufficient? Would you rather see any changes?

File: pkg/scheduler/algorithmprovider/defaults/defaults.go (hunk 7 above)
Comment (damemi) [reply pkg/scheduler/algorithmprovider/defaults/defaults.go]: Didn't know if this was intentional or not, but it seems it wasn't. @ahg-g updated, thanks

**Comment by ahg-g** [comment]
Thanks Mike, I am assuming you amended the commit instead of creating a new one (unless I am somehow missing something on my github UI :)), this makes it hard to see what changes you made compared to the last review, so in the future can you please create new commits instead of amending.

**Comment by ahg-g** [comment]
lgtm

File: pkg/scheduler/algorithmprovider/defaults/defaults.go (hunk 7 above)
Comment (ravisantoshgudimetla) [reply pkg/scheduler/algorithmprovider/defaults/defaults.go]: Ahh sorry for the delay, @damemi pinged me today to look into it. Now, I realized that at that time, we were thinking of deprecating that Predicate altogether, so I did not add it but we did not go that route, so yes, it is needed

https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/algorithm/predicates/predicates.go#L1527

File: pkg/scheduler/api/validation/validation.go (hunk 6 above)
Comment (liggitt) [reply pkg/scheduler/api/validation/validation.go]: one last question... I'm trying to understand the order and dependency between scheduler config (via flags or config file), scheduler policy config (via file), and how the feature flags plumbing affects our ability to evaluate if a given config is valid before making it effective.

1. feature flags come from CLI flags (today), but as scheduler config file work progresses, I'd expect them to be able to be specified in the config file
2. the scheduler config file can reference a policy file
3. part of validating the scheduler config would be to load and validate the policy as well

We want to be able to validate config file and policy file before making them effective. That means all validation functions (and the things that compute the inputs to those functions) must be able to be evaluated without affecting the global effective state.

The primary place I see that needs changing is we have to be able to compute the inputs to validation without changing global state. That means we need a function that computes what ApplyFeatureGates() is returning that takes a featuregate as an argument (doesn't use the global `utilfeature.DefaultFeatureGate`) and returns computed values without modifying the global factory. That way, to validate a config prior to making it effective, we would:
1. load the scheduler config file
2. apply the feature gates specified in the config file to `temporaryFeatureGateSet`
3. call `ComputeFeatureGates(temporaryFeatureGateSet)` (which would have no side effects)
4. validate the policy config passing in the computed feature gates

File: pkg/scheduler/algorithmprovider/defaults/defaults.go [hunk 10]
```diff
@@ -94,6 +112,7 @@ func ApplyFeatureGates() {
 		// Register the priority function to specific provider too.
 		factory.InsertPriorityKeyToAlgorithmProviderMap(factory.RegisterPriorityFunction2(priorities.ResourceLimitsPriority, priorities.ResourceLimitsPriorityMap, nil, 1))
 	}
+	return featureDependency
```
Comment (Huang-Wei) [inline pkg/scheduler/algorithmprovider/defaults/defaults.go]: `featureDependency` is sort of a stable thing. I do believe we should keep it as somewhere as (const) variable, rather than calculating it everytime in `ApplyFeatureGates()` - remember that `ApplyFeatureGates()` may be called in integration tests, and we don't want to do the calculation multiple times.

If we follow this idea, `ApplyFeatureGates()` should read from `featureDependency` and register/unregister predKey/predFunc and prioKey/prioFunc properly.

File: pkg/scheduler/api/validation/validation.go (hunk 6 above)
Comment (Huang-Wei) [reply pkg/scheduler/api/validation/validation.go]: I agree with @liggitt that we shouldn't calculate feature dependencies in ApplyFeatureGates() - see https://github.com/kubernetes/kubernetes/pull/79426/files#r302785224

---

More thoughts on the underlying configuration issue:

Essentially how default scheduler behaves depends on (a) feature gates plus (b) predicates/priorities, the combinations could be:

1. feature gates + default policy (i.e. default predicates/priorities)
1. feature gates + custom policy (i.e. custom predicates/priorities)

Let's focus on the 2nd case here. The following cases we should give validation error:

- custom policy violates (non-GA) feature gates
	This can break the function of the whole kubernetes cluster. E.g. an alpha feature functions across apiserver and scheduler, if the related predicate and/or priority is specified in custom policy file, it may work as the feature gate is disabled in apiserver side. (NOTE: we don't think it's a violation if a feature gate is enabled, but its predicate/priority is missing)
- custom policy violates predicate/priority dependencies
	For example, a predicate is specified in policy file, but its dependent predicate/priority isn't specified, or its violated predicate/priority is specified.
- custom policy violates generic mandatory predicate/priority
	That is to say, there may be some generic mandatory predicate/priority, if it's missing (from custom policy file), scheduler may not work. Right now I didn't see any kind of pred/prio falls into this category, but I may be wrong, or in the future we may have some.

Based on the above conditions, I'd suppose:

1. Maintain a data structure to have mappings from:
	1. feature gate => predicate/priority
	1. predicate/priority => dependent predicate/priority and violated predicate/priority
1. The mapping structure should be static (for a single release), so put it somewhere as a simple `var`
1. Per `scheduler config validation`, do NOT run `ApplyFeatureGates()`; instead get the mapping obj, and ensure all violations are captured.
1. (I will propose another PR) Based on the mapping structure, revise ApplyFeatureGates() - this function looks buggy, esp. when used in integration tests.

@damemi, thoughts?

Comment (Huang-Wei) [reply pkg/scheduler/api/validation/validation.go]: And (in another PR), we should discuss whether we should or should not apply `ApplyFeatureGates()` upon a custom policy config file. If we did, it seems we **mutates** the predicates/priorities collection given by user.

**Comment by damemi** [comment]
@Huang-Wei I pushed a new commit that reverts the changes which made `ValidatePolicy` dependent on a list of feature dependencies and took the calculation of those dependencies out of `ApplyFeatureGates` and put them in a var. Please let me know if I understood you correctly. Also, would you prefer to merge https://github.com/kubernetes/kubernetes/pull/80144 before this?

@liggitt the update I made just loops through all of the relevant feature gates and still uses `DefaultFeatureGate` to determine if they're enabled. I know that the feature gates are currently set by a flag, but could you help me find how to access that flag in the scheduler code? I tried looking through all the flagsets that are added in `NewSchedulerCommand` and its parents but couldn't find it. I imagine it's probably just something like `o.FeatureGates`, right? Maybe I missed it

File: cmd/kube-scheduler/app/server.go [hunk 11]
```diff
@@ -21,6 +21,7 @@ import (
 	"context"
 	"fmt"
 	"io"
+	"k8s.io/kubernetes/pkg/scheduler/algorithmprovider"
```
Comment (Huang-Wei) [inline cmd/kube-scheduler/app/server.go]: imports are grouped, so please revert this change.

File: pkg/scheduler/algorithmprovider/defaults/defaults.go [hunk 12]
```diff
@@ -17,15 +17,13 @@ limitations under the License.
 package defaults
 
 import (
-	"k8s.io/klog"
-
 	"k8s.io/apimachinery/pkg/util/sets"
 	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/klog"
```
Comment (Huang-Wei) [inline pkg/scheduler/algorithmprovider/defaults/defaults.go]: imports are grouped, so please revert this change.

File: pkg/scheduler/algorithmprovider/defaults/defaults.go [hunk 13]
```diff
@@ -83,8 +82,7 @@ func ApplyFeatureGates() {
 		// if you just want insert to specific provider, call func InsertPredicateKeyToAlgoProvider()
 		factory.InsertPredicateKeyToAlgorithmProviderMap(predicates.PodToleratesNodeTaintsPred)
 		factory.InsertPredicateKeyToAlgorithmProviderMap(predicates.CheckNodeUnschedulablePred)
-
-		klog.Infof("TaintNodesByCondition is enabled, PodToleratesNodeTaints predicate is mandatory")
+		klog.Warningf("TaintNodesByCondition is enabled, PodToleratesNodeTaints predicate is mandatory")
```
Comment (Huang-Wei) [inline pkg/scheduler/algorithmprovider/defaults/defaults.go]: why warning?

File: pkg/scheduler/algorithmprovider/defaults/defaults.go [hunk 14]
```diff
@@ -56,7 +55,7 @@ func defaultPredicates() sets.String {
 	)
 }
 
-// ApplyFeatureGates applies algorithm by feature gates.
+// ApplyFeatureGates applies algorithm by feature gates and returns a list of feature dependencies.
```
Comment (Huang-Wei) [inline pkg/scheduler/algorithmprovider/defaults/defaults.go]: please revert the comment

File: cmd/kube-scheduler/app/server.go [hunk 15]
```diff
@@ -169,6 +165,10 @@ func Run(cc schedulerserverconfig.CompletedConfig, stopCh <-chan struct{}, regis
 		}
 	}
 
+	// Apply algorithms based on feature gates.
+	// TODO: make configurable?
+	algorithmprovider.ApplyFeatureGates()
```
Comment (Huang-Wei) [inline cmd/kube-scheduler/app/server.go]: Right now we don't need to move this call, right?

File: pkg/scheduler/api/validation/validation.go [hunk 16]
```diff
@@ -19,19 +19,27 @@ package validation
 import (
 	"errors"
 	"fmt"
+	"k8s.io/component-base/featuregate"
```
Comment (Huang-Wei) [inline pkg/scheduler/api/validation/validation.go]: These imports should be grouped properly. Otherwise it won't pass the CI. I guess you can run hack/update-gofmt.sh to correct these errors, or in your editor configure an instant `gofmt -w $FILE` action upon saving.

Comment (Huang-Wei) [reply pkg/scheduler/api/validation/validation.go]: Another thing is it seems `k8s.io/component-base` is a new dependency, we'd better avoid it unless it's mandatory.

File: pkg/scheduler/api/validation/validation.go [hunk 17]
```diff
@@ -65,6 +73,64 @@ func ValidatePolicy(policy schedulerapi.Policy) error {
...
+		ExcludedPriorityList: nil,
+	},
+}
+
+// validateFeatureDependencies checks dependent and excluded predicates/priorities for given feature gates
+func validateFeatureDependencies(policy schedulerapi.Policy) error {
+	var validationErrors []error
+	for feature, featureDependencies := range FeatureGateDependencies {
+		if !utilfeature.DefaultFeatureGate.Enabled(feature) {
+			continue
```
Comment (Huang-Wei) [inline pkg/scheduler/api/validation/validation.go]: When a featuregate is **disabled**, it'd be good to check the **absence** of its dependents. For instance, if the featuregate `TaintNodesByCondition` is disabled, but the predicate is present in the custom policy, it won't work properly, so we should give a validation error.

Additionally, I'm proposing to separate dependents into a) the predicate/priority itself plus b) its (real) dependents. So that we can only check a) instead of a)+b).
It's because we **may** come across some feature which has its own predicate/priority, and also other ones as its dependents. In that case, we shouldn't check the absence of the dependent ones but just its own one.

Thoughts?

Comment (damemi) [reply pkg/scheduler/api/validation/validation.go]: It would be helpful for the user to check when a feature gate is disabled but its predicate is present, in the case that they forgot to enable the feature, but I think the more important case here is when a feature is enabled and conflicting (excluded) predicates are also enabled. In the first case we are simply reminding the user of something they forgot to enable, which is fine and helpful but not necessarily breaking whereas in the second case an invalid predicate/priority will at least create errors when we remove them in ApplyFeatureGates. Maybe we could check both?

Also, I'm still not sure exactly what you mean in separating the predicate itself plus its dependents. Could you provide an example struct to demonstrate?

File: pkg/scheduler/api/validation/validation.go (hunk 16 above)
Comment (damemi) [reply pkg/scheduler/api/validation/validation.go]: In order to use `utilfeature.DefaultFeatureGate.Enabled()` it expects a `featuregate.Feature`, which isn't currently imported anywhere. I would like to just check against what is enabled in the `--feature-gates` flag as @liggitt suggested (then I won't need this import and can just use `string`s) but I can't find how to access that flag and perhaps pass it to the componentconfig

File: cmd/kube-scheduler/app/server.go (hunk 15 above)
Comment (damemi) [reply cmd/kube-scheduler/app/server.go]: Got feedback from @ravisantoshgudimetla to move it here https://github.com/kubernetes/kubernetes/pull/79426#discussion_r298661507

**Comment by damemi** [comment]
@Huang-Wei addressed some of your feedback (lots of misplaced imports, thanks GoLand) but still have a couple questions that I added inline

**Comment by damemi** [comment]
I believe this is no longer relevant with the switch to plugins, and the removal of these policy maps. We have worked on separate methods to provide similar validation for plugin configs anyway, so I'm going to close this.

If I'm wrong, please reopen and I will pick this back up
